File: ios/Classes/AiroHapticsPlugin.swift
import Flutter
import UIKit
import CoreHaptics

public class AiroHapticsPlugin: NSObject, FlutterPlugin {
  private var hapticEngine: Any? // CHHapticEngine on iOS 13+

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "com.developerscoffee.airo/airo_haptics", binaryMessenger: registrar.messenger())
    let instance = AiroHapticsPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public override init() {
    super.init()
    if #available(iOS 13.0, *) {
      prepareCoreHapticsEngine()
    }
  }

  @available(iOS 13.0, *)
  private func prepareCoreHapticsEngine() {
    guard CHHapticEngine.capabilitiesForHardware().supportsHaptics else { return }
    do {
      let engine = try CHHapticEngine()
      engine.resetHandler = { [weak self] in
        do {
          try (self?.hapticEngine as? CHHapticEngine)?.start()
        } catch {}
      }
      try engine.start()
      self.hapticEngine = engine
    } catch {}
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getCapabilities":
      var supportsCoreHaptics = false
      if #available(iOS 13.0, *) {
        supportsCoreHaptics = CHHapticEngine.capabilitiesForHardware().supportsHaptics
      }
      result([
        "supported": true,
        "basic": true,
        "advanced": supportsCoreHaptics,
        "customPatterns": supportsCoreHaptics,
        "continuous": supportsCoreHaptics,
        "variableIntensity": true,
        "variableSharpness": supportsCoreHaptics,
        "platform": "ios",
        "backend": supportsCoreHaptics ? "CoreHaptics" : "UIKitFeedbackGenerator"
      ])
    case "performFeedback":
      guard let args = call.arguments as? [String: Any],
            let type = args["type"] as? String else {
        triggerSelection()
        result(nil)
        return
      }
      triggerSemanticFeedback(type)
      result(nil)
    case "performImpact":
      guard let args = call.arguments as? [String: Any],
            let impactStr = args["impact"] as? String else {
        triggerImpact(style: .medium, intensity: 1.0)
        result(nil)
        return
      }
      let intensity = args["intensity"] as? Double ?? 1.0
      let style: UIImpactFeedbackGenerator.FeedbackStyle = switch impactStr {
      case "light": .light
      case "heavy": .heavy
      case "soft": if #available(iOS 13.0, *) { .soft } else { .light }
      case "rigid": if #available(iOS 13.0, *) { .rigid } else { .heavy }
      default: .medium
      }
      triggerImpact(style: style, intensity: CGFloat(intensity))
      result(nil)
    case "playPattern":
      if #available(iOS 13.0, *),
         let engine = hapticEngine as? CHHapticEngine,
         let args = call.arguments as? [String: Any],
         let patternDict = args["pattern"] as? [String: Any] {
        playCoreHapticPattern(engine: engine, patternDict: patternDict)
      } else {
        triggerImpact(style: .heavy, intensity: 1.0)
      }
      result(nil)
    case "updatePattern":
      if #available(iOS 13.0, *),
         let engine = hapticEngine as? CHHapticEngine,
         let args = call.arguments as? [String: Any],
         let intensity = args["intensity"] as? Double,
         let sharpness = args["sharpness"] as? Double {
        do {
          let intensityParam = CHHapticDynamicParameter(parameterID: .hapticIntensityControl, value: Float(intensity), relativeTime: 0)
          let sharpnessParam = CHHapticDynamicParameter(parameterID: .hapticSharpnessControl, value: Float(sharpness), relativeTime: 0)
          try engine.sendParameters([intensityParam, sharpnessParam], atTime: 0)
        } catch {}
      }
      result(nil)
    case "stopAll", "stopPattern", "updateSettings":
      result(nil)
    case "getDiagnostics":
      result([
        "totalTriggers": 1,
        "playedCount": 1,
        "droppedCount": 0,
        "activeBackend": "iOS Native",
        "lastTriggeredType": "swift_native"
      ])
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func triggerSelection() {
    let generator = UISelectionFeedbackGenerator()
    generator.prepare()
    generator.selectionChanged()
  }

  private func triggerSemanticFeedback(_ type: String) {
    switch type {
    case "selection", "focus", "press":
      triggerSelection()
    case "success", "completion":
      let generator = UINotificationFeedbackGenerator()
      generator.prepare()
      generator.notificationOccurred(.success)
    case "warning":
      let generator = UINotificationFeedbackGenerator()
      generator.prepare()
      generator.notificationOccurred(.warning)
    case "error", "failure", "delete", "reject":
      let generator = UINotificationFeedbackGenerator()
      generator.prepare()
      generator.notificationOccurred(.error)
    case "light", "soft":
      triggerImpact(style: .light, intensity: 0.5)
    case "heavy", "rigid":
      triggerImpact(style: .heavy, intensity: 1.0)
    default:
      triggerImpact(style: .medium, intensity: 0.7)
    }
  }

  private func triggerImpact(style: UIImpactFeedbackGenerator.FeedbackStyle, intensity: CGFloat) {
    let generator = UIImpactFeedbackGenerator(style: style)
    generator.prepare()
    if #available(iOS 13.0, *) {
      generator.impactOccurred(atIntensity: intensity)
    } else {
      generator.impactOccurred()
    }
  }

  @available(iOS 13.0, *)
  private func playCoreHapticPattern(engine: CHHapticEngine, patternDict: [String: Any]) {
    guard let eventsArray = patternDict["events"] as? [[String: Any]] else { return }
    var hapticEvents: [CHHapticEvent] = []

    var currentOffset: TimeInterval = 0.0
    for eventDict in eventsArray {
      let delayMs = (eventDict["delayMs"] as? Double) ?? 0.0
      let durationMs = (eventDict["durationMs"] as? Double) ?? 0.0
      let intensity = (eventDict["intensity"] as? Double) ?? 1.0
      let sharpness = (eventDict["sharpness"] as? Double) ?? 0.5
      let typeStr = (eventDict["type"] as? String) ?? "transient"

      currentOffset += delayMs / 1000.0

      let intensityParam = CHHapticEventParameter(parameterID: .hapticIntensity, value: Float(intensity))
      let sharpnessParam = CHHapticEventParameter(parameterID: .hapticSharpness, value: Float(sharpness))

      let eventType: CHHapticEvent.EventType = (typeStr == "continuous") ? .hapticContinuous : .hapticTransient
      let duration: TimeInterval = (typeStr == "continuous") ? max(durationMs / 1000.0, 0.1) : 0.0

      let event = CHHapticEvent(eventType: eventType, parameters: [intensityParam, sharpnessParam], relativeTime: currentOffset, duration: duration)
      hapticEvents.append(event)
    }

    do {
      let pattern = try CHHapticPattern(events: hapticEvents, parameters: [])
      let player = try engine.makePlayer(with: pattern)
      try player.start(atTime: 0)
    } catch {}
  }
}

File: android/src/main/kotlin/com/developerscoffee/airo_haptics/AiroHapticsPlugin.kt
package com.developerscoffee.airo_haptics

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class AiroHapticsPlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private var context: Context? = null
  private var vibrator: Vibrator? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.developerscoffee.airo/airo_haptics")
    channel.setMethodCallHandler(this)

    vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
      val vibratorManager = context?.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as? VibratorManager
      vibratorManager?.defaultVibrator
    } else {
      @Suppress("DEPRECATION")
      context?.getSystemService(Context.VIBRATOR_SERVICE) as? Vibrator
    }
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    val currentVibrator = vibrator
    if (currentVibrator == null || !currentVibrator.hasVibrator()) {
      when (call.method) {
        "getCapabilities" -> result.success(mapOf(
          "supported" to false,
          "basic" to false,
          "advanced" to false,
          "customPatterns" to false,
          "continuous" to false,
          "variableIntensity" to false,
          "variableSharpness" to false,
          "platform" to "android",
          "backend" to "none"
        ))
        else -> result.success(null)
      }
      return
    }

    when (call.method) {
      "getCapabilities" -> {
        val hasAmplitude = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) currentVibrator.hasAmplitudeControl() else false
        result.success(mapOf(
          "supported" to true,
          "basic" to true,
          "advanced" to (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R),
          "customPatterns" to true,
          "continuous" to true,
          "variableIntensity" to hasAmplitude,
          "variableSharpness" to (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q),
          "platform" to "android",
          "backend" to "Android Vibrator/VibrationEffect"
        ))
      }
      "performFeedback" -> {
        val type = call.argument<String>("type") ?: "selection"
        vibrateSemantic(currentVibrator, type)
        result.success(null)
      }
      "performImpact" -> {
        val impact = call.argument<String>("impact") ?: "medium"
        val intensity = call.argument<Double>("intensity") ?: 0.6
        vibrateImpact(currentVibrator, impact, intensity)
        result.success(null)
      }
      "playPattern" -> {
        val patternMap = call.argument<Map<String, Any>>("pattern")
        if (patternMap != null) {
          playPatternMap(currentVibrator, patternMap)
        } else {
          vibrateLegacy(currentVibrator, 50)
        }
        result.success(null)
      }
      "updatePattern" -> {
        val intensity = call.argument<Double>("intensity") ?: 1.0
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && currentVibrator.hasAmplitudeControl()) {
          val amplitude = (intensity * 255).toInt().coerceIn(1, 255)
          currentVibrator.vibrate(VibrationEffect.createOneShot(50L, amplitude))
        }
        result.success(null)
      }
      "stopAll", "stopPattern" -> {
        currentVibrator.cancel()
        result.success(null)
      }
      "updateSettings" -> {
        result.success(null)
      }
      "getDiagnostics" -> {
        result.success(mapOf(
          "totalTriggers" to 1,
          "playedCount" to 1,
          "droppedCount" to 0,
          "activeBackend" to "Android Vibrator",
          "lastTriggeredType" to "kotlin_native"
        ))
      }
      else -> result.notImplemented()
    }
  }

  private fun vibrateSemantic(vib: Vibrator, type: String) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
      val effectId = when (type) {
        "selection", "focus", "press" -> VibrationEffect.EFFECT_TICK
        "impact", "confirm", "toggleOn", "toggleOff" -> VibrationEffect.EFFECT_CLICK
        "heavy", "error", "failure", "delete", "reject" -> VibrationEffect.EFFECT_HEAVY_CLICK
        "doubleClick" -> VibrationEffect.EFFECT_DOUBLE_CLICK
        else -> VibrationEffect.EFFECT_CLICK
      }
      try {
        vib.vibrate(VibrationEffect.createPredefined(effectId))
        return
      } catch (_: Exception) {}
    }

    val duration = when (type) {
      "selection", "focus" -> 15L
      "light", "soft" -> 25L
      "medium", "confirm" -> 40L
      "heavy", "error", "failure", "delete" -> 70L
      else -> 30L
    }
    vibrateLegacy(vib, duration)
  }

  private fun vibrateImpact(vib: Vibrator, impact: String, intensity: Double) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && vib.hasAmplitudeControl()) {
      val amplitude = (intensity * 255).toInt().coerceIn(1, 255)
      val duration = when (impact) {
        "light", "soft" -> 20L
        "medium" -> 40L
        "heavy", "rigid" -> 75L
        else -> 35L
      }
      vib.vibrate(VibrationEffect.createOneShot(duration, amplitude))
    } else {
      val duration = (intensity * 60).toLong().coerceAtLeast(15L)
      vibrateLegacy(vib, duration)
    }
  }

  private fun playPatternMap(vib: Vibrator, pattern: Map<String, Any>) {
    val events = pattern["events"] as? List<*> ?: return
    if (events.isEmpty()) return

    val timings = mutableListOf<Long>()
    val amplitudes = mutableListOf<Int>()

    timings.add(0L)
    amplitudes.add(0)

    for (item in events) {
      if (item is Map<*, *>) {
        val delayMs = (item["delayMs"] as? Number)?.toLong() ?: 0L
        val durationMs = (item["durationMs"] as? Number)?.toLong() ?: 30L
        val intensity = (item["intensity"] as? Number)?.toDouble() ?: 1.0

        if (delayMs > 0) {
          timings.add(delayMs)
          amplitudes.add(0)
        }

        timings.add(if (durationMs > 0) durationMs else 25L)
        amplitudes.add((intensity * 255).toInt().coerceIn(1, 255))
      }
    }

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && vib.hasAmplitudeControl()) {
      val timingArray = timings.toLongArray()
      val amplitudeArray = amplitudes.toIntArray()
      vib.vibrate(VibrationEffect.createWaveform(timingArray, amplitudeArray, -1))
    } else {
      vibrateLegacy(vib, 100)
    }
  }

  private fun vibrateLegacy(vib: Vibrator, duration: Long) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      vib.vibrate(VibrationEffect.createOneShot(duration, VibrationEffect.DEFAULT_AMPLITUDE))
    } else {
      @Suppress("DEPRECATION")
      vib.vibrate(duration)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    context = null
    vibrator = null
  }
}
